File: app/src/main/java/com/selflearning/chemistree/adapter/ReactionAdapter.java
package com.selflearning.chemistree.adapter;

import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.selflearning.chemistree.R;
import com.selflearning.chemistree.chemistry.reactions.ReactionEntry;

import java.util.List;

public class ReactionAdapter extends RecyclerView.Adapter<ReactionAdapter.ReactionViewHolder> {

    private List<ReactionEntry> reactionEntryList;
    private String react;

    public ReactionAdapter(List<ReactionEntry> reactionEntryList) {
        this.reactionEntryList = reactionEntryList;
    }

    @NonNull
    @Override
    public ReactionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.rv_reactions, parent, false);
        return new ReactionViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ReactionViewHolder holder, int position) {
        if(reactionEntryList != null){
            ReactionEntry reaction = reactionEntryList.get(position);
            react = reaction.reaction.replaceAll("([a-zA-Z])([0-9]*)", "$1<sub><small>$2</small></sub>");
            holder.tvReaction.setText(Html.fromHtml(react));
        }
    }

    @Override
    public int getItemCount() {
        return reactionEntryList.size();
    }

    class ReactionViewHolder extends RecyclerView.ViewHolder {

        private TextView tvReaction;

        public ReactionViewHolder(@NonNull View itemView) {
            super(itemView);
            tvReaction = itemView.findViewById(R.id.tvReaction);
        }
    }
}

File: app/src/main/java/com/selflearning/chemistree/adapter/ReactionAdapter.kt
package com.selflearning.chemistree.adapter

import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.selflearning.chemistree.R
import com.selflearning.chemistree.adapter.ReactionAdapter.ReactionViewHolder
import com.selflearning.chemistree.chemistry.reactions.ReactionEntry
import com.selflearning.chemistree.databinding.RvReactionsBinding

class ReactionAdapter(private val reactionEntryList: List<ReactionEntry>?) : RecyclerView.Adapter<ReactionViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReactionViewHolder {
        val binding = RvReactionsBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ReactionViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ReactionViewHolder, position: Int) {
        if (reactionEntryList != null) {
            val reaction: ReactionEntry = reactionEntryList[position]
            holder.bind(reaction)
        }
    }

    override fun getItemCount(): Int {
        return reactionEntryList!!.size
    }

    inner class ReactionViewHolder(private val binding: RvReactionsBinding) : RecyclerView.ViewHolder(binding.root) {

        fun bind(reaction: ReactionEntry){
            val react = reaction.reaction.replace("([a-zA-Z])([0-9]*)".toRegex(), "$1<sub><small>$2</small></sub>")
            binding.tvReaction.text = Html.fromHtml(react)
        }
    }
}